rbac/management/group/relation_api_dual_write_group_handler.py

```diff
@@ -123,39 +130,65 @@ def _replicate(self):
+
+    def replicate_added_roles(self, roles: list[Role], remove_default_access_from: Optional[TenantMapping] = None):
+        """Generate group relations and binding mapping for role."""
         if not self.replication_enabled():
             return
 
         def add_group_to_binding(mapping: BindingMapping):
             self.group_relations_to_add.append(mapping.add_group_to_bindings(str(self.group.uuid)))
 
-        def create_default_mapping():
-            assert role.system is True, "Expected system role. Mappings for custom roles must already be created."
-            binding = V2rolebinding(
-                str(uuid4()),
-                # Assumes same role UUID for V2 system role equivalent.
-                V2role.for_system_role(str(role.uuid)),
-                V2boundresource(("rbac", "workspace"), str(self.default_workspace.id)),
-                groups=frozenset([str(self.group.uuid)]),
+        for role in roles:
+
+            def create_default_mapping():
+                assert role.system is True, "Expected system role. Mappings for custom roles must already be created."
+                binding = V2rolebinding(
+                    str(uuid4()),
+                    # Assumes same role UUID for V2 system role equivalent.
+                    V2role.for_system_role(str(role.uuid)),
+                    V2boundresource(("rbac", "workspace"), str(self.default_workspace.id)),
+                    groups=frozenset([str(self.group.uuid)]),
+                )
+                mapping = BindingMapping.for_role_binding(binding, role)
+                self.group_relations_to_add.extend(mapping.as_tuples())
+                return mapping
+
+            self._update_mapping_for_role(
+                role,
+                update_mapping=add_group_to_binding,
+                create_default_mapping_for_system_role=create_default_mapping,
             )
-            mapping = BindingMapping.for_role_binding(binding, role)
-            self.group_relations_to_add.extend(mapping.as_tuples())
-            return mapping
 
-        self._update_mapping_for_role(
-            role, update_mapping=add_group_to_binding, create_default_mapping_for_system_role=create_default_mapping
-        )
+        if remove_default_access_from is not None:
+            default_bindings = self._create_default_relation_tuples(
+                self.default_workspace.id,
+                remove_default_access_from.default_role_binding_uuid,
+                remove_default_access_from.default_group_uuid,
+            )
+            self.group_relations_to_remove.extend(default_bindings)
+
+    def replicate(self):
+        """Replicate added role."""
+        if not self.replication_enabled():
+            return
+
+        if self.group.tenant.tenant_name == "public":
+            return
         self._replicate()
 
-    def replicate_removed_role(self, role: Role):
+    def replicate_removed_roles(self, roles: list[Role]):
```

Comment:
Not strictly needed but if we change the other method it would be consistent
```suggestion
    def replicate_removed_roles(self, roles: Iterable[Role]):
```